### ISSUE TITLE
BUG: some (not all) corner-joined polygons are lost after applying either the dissolve() or the buffer(0) methods

### ISSUE THREAD

**Issue by swiss-knight** [geopandas/geopandas#2008]
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of geopandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of geopandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import rasterio # version 1.2.6
from rasterio.features import shapes
import geopandas as gpd # version 0.9.0
BASEDIR = '/path/to/folder/where/original_raster_is/' # change me!
SRS = 2056 # or whatever SRS you'd like

# Simulate a random binary raster:
N = np.random.randint(2, size=(64,64))
N[50:,54:] = -99999
N = N.astype(np.float32)

results = ({'properties': {'raster_value': v}, 'geometry': s}
    for i, (s,v) in enumerate(shapes(N))
)

geometries = list(results)
gdf = gpd.GeoDataFrame.from_features(geometries)
gdf.set_crs(crs=f'epsg:{SRS}',inplace=True)
gdf.to_file(os.path.join(BASEDIR,'gdf.gpkg'), driver='GPKG') # image 1 - original  gdf
# %%
# Up to this point, everything is perfectly fine

exported_1 = gdf[gdf['raster_value']==1] # export only polygons which have a value of 1
exported_1.to_file(os.path.join(BASEDIR,'exported_1.gpkg'), driver='GPKG') # image 2 - exported_1

# %%
# From here, there are invalid geometries:
exported_1.is_valid.all() # False
invalids = exported_1[~exported_1.is_valid]
idx = exported_1[~exported_1.is_valid].index

# If I want to dissolve the exported_1 geodataframe:
dissolved = exported_1.dissolve()
dissolved.is_valid.all() # True
dissolved.to_file(os.path.join(BASEDIR,'dissolved.gpkg'), driver='GPKG') # image 3 - dissolved
# There is a problem, some polygons are missing... but there are no more invalid geometries. 

# Let's also try the .buffer(0) trick just in case:
cleaned = exported_1.buffer(0) # this cleans the invalid geometries
cleaned.is_valid.all() # True

cleaned.to_file(os.path.join(BASEDIR,'cleaned.gpkg'), driver='GPKG') # image 4 - cleaned
# Same problem: some polygons are missing
```

#### Problem description
I have raster file (GeoTIFF) having values of either 0 or 1. This is the result of a viewshed analysis from an observer point relative to a DEM. 1 means the cell of the DEM is "visible", and 0 means it's "invisible" from the observer point.   
In order to run some vector computations, for storage and sharing reasons, I have vectorized this result using `rasterio`.
I only need to keep polygons which depict visible regions (value of 1) and I need them all as a single multipolygon, so I have to dissolve them.

When I export the original `gdf` every polygons are present.
When I export the `dissolved` or the `cleaned` geodataframe; some polygons are lost. They are all "corner-joined" black regions, which disappeared, probably "closed" (morphologicaly speaking) by either the dissolve operation or the buffer(0) (both methods seems to give the exact same result), as if they were randomly picking one of the two shape sharing a common corner.... This apparent randomness may be related to the 'orientation' of such pieces if that makes sense.

More detailed info here: https://gis.stackexchange.com/questions/403523/wrong-features-when-exporting-vectorized-raster-data-to-a-file-using-geopandas-a 

image 1 - original `gdf`:
![image](https://user-images.githubusercontent.com/19967599/125696353-c40f6425-1692-4f58-8b93-d92aef2d55aa.png)

Image 2 - `exported_1`:
![image](https://user-images.githubusercontent.com/19967599/125696448-87c8485c-474f-47de-ad00-4fc2f8f26075.png)

Image 3 - `dissolved`:
![image](https://user-images.githubusercontent.com/19967599/125696468-474bdcc0-ff40-4565-860e-3a7cf9ce8819.png)

Image 4 - `cleaned`:
![image](https://user-images.githubusercontent.com/19967599/125696475-b7cfd58f-faa5-4524-94ea-b250fc836a77.png)

Image 5 - adding transparency on the `dissolved`gdf  with the original `gdf` in the background makes the missing polygons appear as darker ([deeper](https://www.youtube.com/watch?v=OorZcOzNcgE)) purple regions:
![image](https://user-images.githubusercontent.com/19967599/125693047-43cfc2ce-6fb0-4b57-8bf4-44244e92702b.png)

All images are screenshots from [QGIS](https://github.com/qgis/QGIS) 3.16 after having loaded the exported `https://github.com/opengeospatial/geopackage`.

#### Expected Output
A result keeping the original shapes in the first `gdf`, which also matches the original raster cells, after either applying `dissolve()` or `buffer(0)`, the former being privileged because I do need a unique multipolygon. 

#### Output of ``geopandas.show_versions()``

<details>

SYSTEM INFO
-----------
python     : 3.6.9 (default, Jan 26 2021, 15:33:00)  [GCC 8.4.0]
executable : /usr/bin/python3.6
machine    : Linux-5.4.0-77-generic-x86_64-with-Ubuntu-18.04-bionic

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : 3.7.1
GEOS lib   : /usr/lib/x86_64-linux-gnu/libgeos_c.so
GDAL       : 3.3.0
GDAL data dir: /usr/local/lib/python3.6/dist-packages/fiona/gdal_data
PROJ       : 7.2.0
PROJ data dir: /usr/local/lib/python3.6/dist-packages/pyproj/proj_dir/share/proj

PYTHON DEPENDENCIES
-------------------
geopandas  : 0.9.0
pandas     : 1.1.5
fiona      : 1.8.20
numpy      : 1.19.5
shapely    : 1.7.1
rtree      : 0.9.3
pyproj     : 3.0.0.post1
matplotlib : 3.3.4
mapclassify: 2.4.2
geopy      : 1.21.0
psycopg2   : 2.9.1 (dt dec pq3 ext lo64)
geoalchemy2: 0.9.2
pyarrow    : None
pygeos     : None

</details>

**Comment by martinfleis** [comment]
Thanks for the report (and moving it here from SO). 

The problem starts in `rasterio`. This is the `gdf` coming from `rasterio.features.shapes` plotted with some transparency. You can see that there are overlapping polygons.
![Screenshot 2021-07-14 at 23 27 34](https://user-images.githubusercontent.com/36797143/125701019-f4ae2aff-9555-42cc-acb9-20e16e70a743.png)

I am not sure how to fix that. It seems that GEOS is having trouble validating those polygons from rasterio. It produces a lot of moments where different parts of a polygon share only a corner, which is known to cause issues.

**Comment by martinfleis** [comment]
Just to add, the behaviour is the same with pygeos 0.10 and GEOS 3.9.1.

**Comment by swiss-knight** [comment]
I explored it more in depth with a super tiny 8x8 image, without NaN.

Here are the geometries (`geoms = list(results)`) corresponding to the following image (origin (0,0) of the coordinates is located at the bottom-left corner)     x-axis: →    
y-axis: ↑    

```
[{'properties': {'raster_value': 0.0},
  'geometry': {'type': 'Polygon',
   'coordinates': [[
     (0.0, 0.0),
     (0.0, 2.0),
     (1.0, 2.0),
     (1.0, 3.0),
     (2.0, 3.0),
     (2.0, 1.0),
     (1.0, 1.0),
     (1.0, 0.0),
     (0.0, 0.0)]]}},
 {'properties': {'raster_value': 1.0},
  'geometry': {'type': 'Polygon',
   'coordinates': [[
     (1.0, 0.0),
     (1.0, 1.0),
     (2.0, 1.0),
     (2.0, 0.0),
     (1.0, 0.0)]]}},
 {'properties': {'raster_value': 0.0},
  'geometry': {'type': 'Polygon',
   'coordinates': [[
     (2.0, 0.0),
     (2.0, 1.0),
     (3.0, 1.0),
     (3.0, 2.0),
     (5.0, 2.0),
     (5.0, 0.0),
     (2.0, 0.0)]]}},
 {'properties': {'raster_value': 1.0},
  'geometry': {'type': 'Polygon',
   'coordinates': [[
     (5.0, 0.0),
     (5.0, 2.0),
     (6.0, 2.0),
     (6.0, 0.0),
     (5.0, 0.0)]]}},
 {'properties': {'raster_value': 1.0},
  'geometry': {'type': 'Polygon',
   'coordinates': [[
     (2.0, 1.0),
     (2.0, 3.0),
     (5.0, 3.0),
     (5.0, 2.0),
     (3.0, 2.0),
     (3.0, 1.0),
     (2.0, 1.0)]]}},
 {'properties': {'raster_value': 1.0},
  'geometry': {'type': 'Polygon',
   'coordinates': [[
     (0.0, 2.0),
     (0.0, 3.0),
     (1.0, 3.0),
     (1.0, 2.0),
     (0.0, 2.0)]]}},
 {'properties': {'raster_value': 0.0},
  'geometry': {'type': 'Polygon',
   'coordinates': [[
     (5.0, 2.0),
     (5.0, 3.0),
     (6.0, 3.0),
     (6.0, 2.0),
     (5.0, 2.0)]]}},
 {'properties': {'raster_value': 0.0},
  'geometry': {'type': 'Polygon',
   'coordinates': [[
     (0.0, 3.0),
     (0.0, 4.0),
     (1.0, 4.0),
     (1.0, 3.0),
     (0.0, 3.0)]]}},
 {'properties': {'raster_value': 1.0},
  'geometry': {'type': 'Polygon',
   'coordinates': [[
     (5.0, 3.0),
     (5.0, 4.0),
     (6.0, 4.0),
     (6.0, 3.0),
     (5.0, 3.0)]]}},
 {'properties': {'raster_value': 1.0},
  'geometry': {'type': 'Polygon',
   'coordinates': [[
     (1.0, 3.0),
     (1.0, 4.0),
     (0.0, 4.0),
     (0.0, 6.0),
     (1.0, 6.0),
     (1.0, 5.0),
     (2.0, 5.0),
     (2.0, 3.0),
     (1.0, 3.0)]]}},
 {'properties': {'raster_value': 1.0},
  'geometry': {'type': 'Polygon',
   'coordinates': [[
     (3.0, 4.0),
     (3.0, 5.0),
     (5.0, 5.0),
     (5.0, 4.0),
     (3.0, 4.0)]]}},
 {'properties': {'raster_value': 0.0},
  'geometry': {'type': 'Polygon',
   'coordinates': [[
     (2.0, 3.0),
     (2.0, 5.0),
     (1.0, 5.0),
     (1.0, 6.0),
     (6.0, 6.0),
     (6.0, 4.0),
     (3.0, 4.0),
     (3.0, 5.0),
     (5.0, 5.0),
     (5.0, 3.0),
     (2.0, 3.0)]]}}]
```

Original raster :
![image](https://user-images.githubusercontent.com/19967599/125766106-27ea01b9-6618-419c-86b8-18cf4a836d7d.png)


GeoDataFrame with `alpha=0.5` :

![image](https://user-images.githubusercontent.com/19967599/125766136-3eb227ba-63ee-42e5-97bf-f928dcef81ce.png)


The "drawing path" of the this last block:
```
 {'properties': {'raster_value': 0.0},
  'geometry': {'type': 'Polygon',
   'coordinates': [[
     (2.0, 3.0),
     (2.0, 5.0),
     (1.0, 5.0),
     (1.0, 6.0),
     (6.0, 6.0),
     (6.0, 4.0),
     (3.0, 4.0),
     (3.0, 5.0),
     (5.0, 5.0),
     (5.0, 3.0),
     (2.0, 3.0)]]}}]
```
is shown here:
![image](https://user-images.githubusercontent.com/19967599/125768322-381751ad-92af-4be2-9efa-233cc48506a4.png)

Vertex 0 is the first and last one defined in that block.      
Please, notice that the red point is actually **not** in the coordinates list. 
In a way, this kind of "flips" the interior ring's topology....    
That's may be an explanation(?) 

While the little blue square flaged 'lb' correspond to that block:
```
 {'properties': {'raster_value': 1.0},
  'geometry': {'type': 'Polygon',
   'coordinates': [[
     (5.0, 3.0),
     (5.0, 4.0),
     (6.0, 4.0),
     (6.0, 3.0),
     (5.0, 3.0)]]}},
```

Both these shapes seems defined in a clockwise manner.     
The red point `(5.0, 4.0)` being defined only in the latter.

**Comment by martinfleis** [comment]
I have a solution.

`pygeos` can fix these geometries.

```py
import pygeos

fixed_geometry = geopandas.GeoSeries(pygeos.make_valid(df.geometry.values.data))
```

Requires GEOS 3.8.0 or newer. Make sure you create your GeoDaraFrame after you install pygeos, that way the `df.geometry.values.data` trick will return array of pygeos geometries.

**Comment by martinfleis** [comment]
If there's something to be fixed, that it is in GEOS which does the operation. They are aware of the issue and working on better tools for fixing such geometries (http://lin-ear-th-inking.blogspot.com/2021/05/fixing-invalid-geometry-with-jts.html).

**Comment by swiss-knight** [comment]
Update:
I've tested what is proposed on the GDAL issue at OSGeo/gdal#1158 and indeed, since a week, with the version 3.3.2 of GDAL the geometry has become valid now (thanks @ GDAL devs).

With `gdal_polygonize.py` from GDAL 3.3.0:

```
Layer name: SELECT
OGRFeature(SELECT):0
  st_isvalidreason(geom) (String) = Self-intersection[2 2]
  POLYGON ((0 4,0 0,4 0,4 4,0 4),(1 3,1 2,3 2,3 1,2 1,2 3,1 3))
```

With `gdal_polygonize.py` from GDAL 3.3.**2**:
```
OGRFeature(SELECT):0
  st_isvalidreason(geom) (String) = Valid Geometry
  POLYGON ((0 4,0 0,4 0,4 4,0 4),(1 3,1 2,2 2,2 3,1 3),(2 2,2 1,3 1,3 2,2 2))
```

**Comment by swiss-knight** [comment]
The issue seems to has been fixed by `rasterio==1.2.8`! (which probably relies on the new GDAL 3.3.2...)

My GeoPandas' version (0.9.0) has not changed in between, but I noticed after having upgraded only `rasterio` from `1.2.6` to `1.2.8` that the geometries are now valid!

\o/
